File: 0232-implement-queue-using-stacks/solution.cpp
class MyQueue {
public:
    stack<int> st;
    stack<int> s;
    MyQueue() {}

    void push(int x) { st.push(x); }

    int pop() {
        while (!st.empty()) {
            s.push(st.top());
            st.pop();
        }
        int x = s.top();
        s.pop();
        while (!s.empty()) {
            st.push(s.top());
            s.pop();
        }
        return x;
    }

    int peek() {
        while (!st.empty()) {
            s.push(st.top());
            st.pop();
        }
        int x = s.top();
        while (!s.empty()) {
            st.push(s.top());
            s.pop();
        }
        return x;
    }

    bool empty() {
        if (st.size() == 0)
            return true;
        return false;
    }
};

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue* obj = new MyQueue();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->peek();
 * bool param_4 = obj->empty();
 */

File: 0232-implement-queue-using-stacks/solution.php
class MyQueue {

    private $s1 = [];
    private $s2 = [];

    function __construct() {
        // initialize stacks
    }

    // push → always s1 me
    function push($x) {
        array_push($this->s1, $x);
    }

    // pop → s2 se
    function pop() {
        if (empty($this->s2)) {
            while (!empty($this->s1)) {
                array_push($this->s2, array_pop($this->s1));
            }
        }
        return array_pop($this->s2);
    }

    // peek → s2 ka top
    function peek() {
        if (empty($this->s2)) {
            while (!empty($this->s1)) {
                array_push($this->s2, array_pop($this->s1));
            }
        }
        return end($this->s2);
    }

    // empty → dono empty hone chahiye
    function empty() {
        return empty($this->s1) && empty($this->s2);
    }
}
